pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -80,14 +80,47 @@ func InitController(kubeClient clientset.Interface, portTable *portcache.PortTab
 		},
 	}
 
-	cacheStore, controller := cache.NewInformer(lw, &corev1.Pod{}, resyncPeriod,
+	podUpdated := func(old, cur interface{}) bool {
+		oldPod, oldok := old.(*corev1.Pod)
+		curPod, curok := cur.(*corev1.Pod)
+		if oldok && curok && oldPod.ResourceVersion != curPod.ResourceVersion {
+			return true
+		}
+		return false
+	}
+
+	podCacheStore, podController := cache.NewInformer(podlw, &corev1.Pod{}, resyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.EnqueuePod,
+			DeleteFunc: c.EnqueuePod,
+			UpdateFunc: func(old, cur interface{}) {
+				if podUpdated(old, cur) {
+					c.EnqueuePod(cur)
+				}
+			},
+		},
+	)
+	c.PodCacheStore = podCacheStore
+	c.PodController = podController
+
+	svclw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return kubeClient.CoreV1().Services(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return kubeClient.CoreV1().Services(metav1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+		},
+	}
+	svcCacheStore, svcController := cache.NewInformer(svclw, &corev1.Service{}, resyncPeriod,
```

Comment:
We should use the `informerFactory` to create a serviceInformer as there's another feature in agent also watching services, otherwise agent would end up with doubling pressure to K8s API and store duplicate resources in memory.

Reply:
would you recommend doing this for watching pods as well, in NPL? cc: @monotosh-avi 
Using informerFactory helps addressing another comment from you
> This won't scall well. We should use the PodLister of PodInformer to avoid more queries.

I'm not sure how to access the informer from the current initialisation. Might have to initialise PodInformer separately to avoid kubeclient calls for Pods.

Reply:
I think I did leave a comment on the initial NPL PR recommending using the shared informer factory to build the informer. However, when it was just Pods that were being watched it wasn't such a big deal as NPL is the only part of the Antrea Agent that watches Pods. I do feel like we should just use it for both Pods & Services now.

Reply:
Originally I thought the Informer doesn't support specifying filter. I just took another look and found it's doable.  
[The informerFactory interface can only set filter for all listers](https://github.com/kubernetes/kubernetes/blob/4b2a98442ea1ffbed9f4335db2a949c3d6a7995e/staging/src/k8s.io/client-go/informers/factory.go#L100-L102), but we could use `NewFilteredPodInformer` to get separate Informer/Indexer for Pod resource, as NPL is still the only one that watches Pods, I think it doesn't matter if it does't use the shared factory to create it. We can revise it if we have another feature that watches Pods in future.